lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+			return -ENOMEM;
+		}
+
+		for_each_feature_param(param, pparam, feature) {
+			entry = allocate_feature_param_entry();
+			if (!entry)
+				return -ENOMEM;
+
+			init_feature_param_entry(entry, cfg, param);
+
+			if (!lisa_fs_create_single(dentry, param->name, NULL,
+						   &lisa_param_feature_fops,
+						   S_IFREG | S_IRUGO, entry)) {
+				free_feature_param_entry(entry);
+				return -ENOMEM;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/////////////////////////////////////
+//           Super block
+/////////////////////////////////////
+
+static struct super_operations lisa_super_ops = {
+	.statfs = simple_statfs,
+};
+
+static int lisa_fs_fill_super(struct super_block *sb, struct fs_context *fc)
+{
+	struct lisa_cfg *cfg;
+	struct inode *root;
+	int ret = -ENOMEM;
+
+	sb->s_maxbytes = MAX_LFS_FILESIZE;
+	sb->s_blocksize = PAGE_SIZE;
+	sb->s_blocksize_bits = PAGE_SHIFT;
+	sb->s_magic = LISA_FS_SUPER_MAGIC;
+	sb->s_op = &lisa_super_ops;
+
+	root = lisa_fs_create_inode(sb, S_IFDIR | S_IRUGO);
+	if (!root)
+		return -ENOMEM;
+
+	root->i_op = &simple_dir_inode_operations;
+	root->i_fop = &simple_dir_operations;
+
+	sb->s_root = d_make_root(root);
+	if (!sb->s_root)
+		goto error1;
+
+	cfg = allocate_lisa_cfg("root");
+	if (!cfg)
+		goto error2;
+
+	mutex_lock(&interface_lock);
+
+	ret = lisa_fs_create_files(sb->s_root, true, cfg);
+	if (ret)
+		goto error3;
+
+	ret = init_lisa_cfg(cfg, &cfg_list, NULL);
+	if (ret)
+		goto error4;
+
+	mutex_unlock(&interface_lock);
+
+	return 0;
+
+error4:
+	free_lisa_cfg(cfg);
+error3:
+	mutex_lock(&interface_lock);
+error2:
+	dput(sb->s_root);
+error1:
+	iput(root);
+
+	return ret;
+}
+
+static int lisa_fs_get_tree(struct fs_context *fc)
+{
+	return get_tree_single(fc, lisa_fs_fill_super);
+}
+
+static const struct fs_context_operations lisa_fs_context_ops = {
+	.get_tree = lisa_fs_get_tree,
+};
+
+static int lisa_init_fs_context(struct fs_context *fc)
+{
+	fc->ops = &lisa_fs_context_ops;
+	put_user_ns(fc->user_ns);
+	fc->user_ns = get_user_ns(&init_user_ns);
+	fc->global = true;
+	return 0;
+}
+
+static void lisa_fs_kill_sb(struct super_block *sb)
+{
+	drain_lisa_cfg(&cfg_list);
+
+	/*
+	 * Free the lisa_features_param param,
+	 * which is not bound to any feature.
+	 */
+	drain_feature_param_entry_value(&lisa_features_param.global_value);
+
+	/* Free the inodes/dentries. */
+	kill_litter_super(sb);
+}
+
+static struct file_system_type lisa_fs_type = {
+	.owner = THIS_MODULE,
+	.name = "lisa",
+	.init_fs_context = lisa_init_fs_context,
+	.kill_sb = lisa_fs_kill_sb,
+};
+
+/*
+ * Note: Cannot initialize global_value list of an unnamed struct in __PARAM
+ * using LIST_HEAD_INIT. Need to have a function to do this.
+ */
+void init_feature_param(void)
+{
+	struct feature_param *param, **pparam;
+	struct feature *feature;
+
+	for_each_feature(feature)
+		for_each_feature_param(param, pparam, feature)
+			INIT_LIST_HEAD(&param->global_value);
+}
+
+int init_lisa_fs(void)
+{
+	int ret;
+
+	init_feature_param();
```

Comment:
Why not making it part of module init instead ? Might be better to distinguish vfs and features.

Reply:
Yes ok